src/client_side_reply.cc

```diff
@@ -906,27 +906,33 @@ purgeEntriesByUrl(HttpRequest * req, const char *url)
 {
 #if USE_HTCP
     bool get_or_head_sent = false;
-    // Optimization: Do not call expensive Root().get() below unless it is needed.
-    const bool needEntry = neighborsHtcpClearNeeded(HTCP_CLR_INVALIDATION);
 #endif
 
     for (HttpRequestMethod m(Http::METHOD_NONE); m != Http::METHOD_ENUM_END; ++m) {
         if (m.respMaybeCacheable()) {
             const cache_key *key = storeKeyPublic(url, m);
             debugs(88, 5, m << ' ' << url << ' ' << storeKeyText(key));
+            if (StoreEntry *entry = Store::Root().get(key))
+            {
 #if USE_HTCP
-            // TODO: Remove if HTCP_CLR_INVALIDATION does not need Store ID (i.e., entry->uri()).
-            if (needEntry) {
-                if (StoreEntry *entry = Store::Root().get(key)) {
-                    entry->lock("purgeEntriesByUrl");
-                    neighborsHtcpClear(entry, url, req, m, HTCP_CLR_INVALIDATION);
-                    if (m == Http::METHOD_GET || m == Http::METHOD_HEAD)
-                        get_or_head_sent = true;
-                    entry->unlock("purgeEntriesByUrl");
+                neighborsHtcpClear(entry, url, req, m, HTCP_CLR_INVALIDATION);
+                if (m == Http::METHOD_GET || m == Http::METHOD_HEAD) {
+                    get_or_head_sent = true;
                 }
-            }
 #endif
-            Store::Root().unlinkByKeyIfFound(key);
+                // Are there local collapsed clients we should notify?
+                // If yes, get transients index and use it later.
+                const auto transientsIndex = entry->hasTransients() && entry->locked() ? entry->mem_obj->xitTable.index : -1;
```

Comment:
I understand that you call `locked()` here because `Controller::get()` might return an unlocked entry, and the `transientsIndex` is not going to survive `StoreEntry::release()` of an unlocked entry. I think we should make this code safer and simpler by always locking/unlocking the entry returned by `get()`. Doing so will remove this complex `locked()` logic and make post-`release()` entry manipulation safe. As a positive side-effect, it may also delete unlocked intransit entries that some other code might leave behind.

```
if (auto entry = Store::Root().get(key)) {
    entry->lock("purgeEntriesByUrl");
    ...
    // entry->release() notifies other waiting workers but
    // XXX: does not abort local collapsed Store clients if needed.
    entry->release(true);
    // Work around the above XXX.
    // TODO: Move into release() after making invokeHandlers() asynchronous.
    if (entry->hasTransients())
        Store::Root().syncCollapsed(entry->mem_obj->xitTable.index);
    entry->unlock("purgeEntriesByUrl");
} else ...
```

Reply:
Done at fdcfaa3.

Reply:
I am afraid I applied your improvement in hurry: I forgot about one reason why I decided to implement without locks. When we get from the Store, we could create a 'fake' entry (when store_table misses it). We need this entry only for notification purpose in release(). The syncCollapsed() method should handle local 'detached' entries, but it should not handle(abort for example) our 'fake' entry! That is why my previous version destroyed this fake entry first(after receiving required index) and then called syncCollapsed().
So I would suggest to undo that fdcfaa3.

Reply:
> When we get from the Store, we could create a 'fake' entry (when store_table misses it). 

Why do you call that entry "fake"? It is a regular/valid entry AFAICT. Does it have some special properties that are important in this context?

> We need this entry only for notification purpose in release().

Yes, and, until that XXX is fixed, for the synchronization purposes in syncCollapsed().

> The syncCollapsed() method should handle local 'detached' entries, 

Yes.

> but it should not handle(abort for example) our 'fake' entry!

Why not?

Reply:
> Why do you call that entry "fake"? 
I call this 'fake' because of my understanding of its usage. It is not used by clients. It is a 'helper' way to notify others. We could avoid using it at all, if, for example, transients->unlinkByKeyIfFound() had a mechanism to notify all workers(including this one) in case Transients has this entry. That was my initial suggestion you did not like.

> Why not?

My suggestion prevents this extra work done on this entry in syncCollapsed(), including probably unnecessary logging. If there is no local entry, we would return after transients->findCollapsed(xitIndex).

Reply:
Thank you for explaining your definition of "fake entry" in this context. I do not think it is a good term for the very real StoreEntry object, obtained using a standard interface, shared with other local clients (if any), and representing very real concurrent transaction(s), but that terminology disagreement does not matter much here as long as we use it consistently.

> Eduard: The syncCollapsed() method should not handle(abort for example) our 'fake' entry!

> Alex: Why not?

> Eduard: My suggestion prevents this extra work done on this entry in syncCollapsed(), including probably unnecessary(and misleading) logging.

I am interpreting your answer as

> The syncCollapsed() method may handle (including aborting) our "fake" entry but, compared to "alternative X",  such handling may produce more debugging and may produce misleading debugging. (where "alternative X" is "unlinkByKeyIfFound() notifies all workers")

It is difficult for me to find all the relevant conversation pieces on GitHub, but, IIRC, "alternative X" was more than just adding same-worker notification. We ended up with the current code by fixing various problems in earlier approaches rather than arbitrary selecting among several simple alternatives. However, we can (and perhaps should) go back to evaluating your earlier suggestion (whatever it was) if the current code is deficient _and_ you believe that your earlier plan would address those deficiencies without introducing worse side effects. Please re-make a specific proposal if you think it would improve the code overall.

For now, AFAICT, the only specific deficiency you have nominated for the current code is "misleading debugging":

> the abort() method says: "Someone wants to abort this transfer." What "transfer" do we have with our "fake" entry? If there is no local entry, we would return after transients->findCollapsed(xitIndex).

* If there is a local writer, then "transfer" refers to that transaction. It should be abort()ed.
* If there is no local writer, then we should not call StoreEntry::abort() and, hence, there will be no misleading debugging.

Do you agree with the above two bullets? If yes, do we need to adjust the code to satisfy them?

I now suspect that you are not applying the term "fake" to any StoreEntry object inside purgeEntriesByUrl(). You are calling _some_ of those objects (i.e., those without local clients) "fake" and then proceed to criticize our handling of those specific objects, even though the code is the same for all objects. In other words, you are saying that "_when_ a given entry happens to be 'fake', we have problems". If so, then that implicit "when" confused me. I hope the above two bullets clarify my position.

Reply:
>-  If there is a local writer, then "transfer" refers to that transaction. It should be abort()ed.
>- If there is no local writer, then we should not call StoreEntry::abort() and, hence, there will be no misleading debugging.

I assume these bullets describe our scenario of received DELETE request and caused purgeEntriesByUrl() activities. I am a bit confused with this description because, IIRC, we discussed that Squid should not abort writers on DELETE: it is up to server. However, still detached _readers_ should be aborted, because they will not get a chance to attach to the already released shared entry (when response finally comes).

Reply:
>  Please re-make a specific proposal if you think it would improve the code overall.

Currently, purgeEntriesByUrl() has release(), followed by syncCollapsed(). This is needed to  properly handle(abort or release) local clients, when required. This approach fixes the problem in purgeEntriesByUrl(), but there can be now(or appear in future) another release() not followed by syncCollapsed().
My suggestion is to make syncCollapsed() work for all local clients without direct call, i.e., do Broadcast() to all workers including current one. This 'full' broadcasting should be done only from Transients::abandon() (e.g., with an additional bool parameter passed to CollapsedForwarding::Broadcast(e)).

Reply:
> Squid should not abort writers on DELETE. However, still detached readers should be aborted.

Yes, you are right. My bullets were wrong!

Besides genuine store writers that do regular abort() calls, Controller has to abort local transients that are waiting for remote writers in vain. Controller is essentially emulating what would happed if those transients were in the aborting writers worker. (The special case of DELETE hitting unattached transients also falls into this pattern, but that is just a hack. Conceptually, Controller is just relaying a remote abort to local transients).

Reply:
> Currently, purgeEntriesByUrl() has release(), followed by syncCollapsed(). This is needed to properly handle(abort or release) local clients, when required. This approach fixes the problem in purgeEntriesByUrl(), 

Agreed. Please note that the explicit/extra syncCollapsed() call hack is needed specifically to work around the lack of local syncCollapsed() effects after unlink()/markForUnlink() (which are called from release()).

> but there can be another release() not followed by syncCollapsed().

Good point! To be more precise, there could be other sources of unlink()/markForUnlink() calls in a non-writing worker. Each such call is a candidate for the same hack.

> My suggestion is to Broadcast() to all workers, including current one. This 'full' broadcasting should be done only from Transients::abandon()

Sounds good. However, please check that the broadcasting queues work for sending notifications to self. I remember that some of the earlier IPC queue designs could not do that, but I do not remember the final state. If broadcast-to-self is a problem, we can do an AsyncCall to Coordinator. That may not be trivial either (because there is probably no suitable Dialer already), but it may be a lot easier than enhancing the mind-boggling IPC queuing code.

And, AFAICT, Transients::markForUnlink() should also broadcast. The comment in that method says that "Controller ensures that this worker has no StoreEntry to abandon() here" but that assurance is not relevant to us. Even without a local StoreEntry, we may need to warn other workers that _their_ collapsed transactions may get stuck after _we_ mark the store entry that they have not attached to yet. Absence of a local StoreEntry is not the same as absence of an entry in the shared transients map... The broadcast from Transients::markForUnlink() can be a regular (i.e., not "full") broadcast because there are no local readers.

Reply:
> And, AFAICT, Transients::markForUnlink() should also broadcast.

s/markForUnlink/unlinkByKeyIfFound/
This unlinkByKeyIfFound() method "allows us to mark being-created
entries (that are locked for writing..", as described in our preamble. If this entry is locked for writing, there should be no attached readers to notify. However a local writer entry (stored in the store_table) will be still available (the corresponding shared Transients entry will be marked for deletion), and it is possible that  a new client, coming on this worker will attach to it.

So, looks like Transients::unlinkByKeyIfFound() should do notification, but probably there is another way to handle this. What if check waitingToBeFreed flag just after we switch from "writing" to "appending"? In other words, if we were marked for deletion during writing to Transients period of time, we should release the local StoreEntry then. For example, we can put this check into Transients::startWriting() just after map->startAppending(index).

Reply:
> If [a Transients] entry is locked for writing, there should be no attached readers to notify.

True, but Transients::unlinkByKeyIfFound() may be called in many situations. Can we guarantee that it is never called in a situation where the Transients entry is _not_ locked for writing?

> However (after switching into "appending" mode), the local writer entry (stored in the store_table) will be still available

I think we must guarantee that Transients::unlinkByKeyIfFound() is not called when there is a local StoreEntry object. Otherwise, we will be going in circles where every method must do everything.

If there is a local StoreEntry, the code may reach Transients::abandon() but must never reach Transients::unlinkByKeyIfFound(). I am not sure, but I suspect that the same or similar invariant holds for all unlinkByKeyIfFound() methods. Unfortunately, ensuring these invariant(s) may overlap with the postponed unlink/markForUnlink polishing, but please see whether you can do something without a significant overhaul.

> What if check waitingToBeFreed flag just after we switch from "writing" to "appending"?

*IIRC*, there was such code somewhere already (but perhaps not in Transients). The idea sounds very familiar... Please see if you can find it.

> we can put this check into Transients::startWriting() just after map->startAppending(index).

Just _before_ startAppending() perhaps? You do not want to let other workers attach to our Transient entry if we are gong to call release() immediately. Again, please see if there was similar code somewhere already. I remember struggling with similar problems earlier.

Reply:
> I think we must guarantee that Transients::unlinkByKeyIfFound() is not called when there is a local StoreEntry object. (1)
> If there is a local StoreEntry, the code may reach Transients::abandon() but must never reach Transients::unlinkByKeyIfFound().(2)

(1): I don't think so. IIRC, one of the Transients::unlinkByKeyIfFound() purposes was to cover situation when we are purging an entry already 'disconnected' from the Transients (i.e., Transients::disconnect() was called). (2) is also problematic because if the local entry is 'disconnected' (does not have Transients index), how abandon() and notification will work without knowing this index?
So, the existing Transients::markForUnlink() code looks correct:

```
if (e.hasTransients())
    abandon(e);
else
    unlinkByKeyIfFound(reinterpret_cast<const cache_key*>(e.key));
```

However this leaves one question: how to notify this 'disconnected' local entry if this is needed (e.g., to mark it for deletion)? The code shows that an entry may become 'disconnected', e.g., after StoreEntry::purgeMem(). It looks like there are situations when it is not 'released', does not have memory object and still lives in the store_table, letting others attach to it.

Reply:
> Alex: I think we must guarantee that Transients::unlinkByKeyIfFound() is not called when there is a local StoreEntry object.

> Eduard:  I don't think so. IIRC, one of the Transients::unlinkByKeyIfFound() purposes was to cover situation when we are purging an entry already 'disconnected' from the Transients (i.e., Transients::disconnect() was called).

> Eduard:  However this leaves one question: how to notify this 'disconnected' local entry?

This is why I am fearing that we are going in circles: We break B while fixing A and then break A again while fixing B. To avoid this endless looping, we need to establish consistent rules/invariants and apply them to every questionable situation in PR scope. The current code may not support any invariant "as is", but it should be possible to eventually adjust the code to support good/correct invariants (i.e., the problem has to have a solution). Thus, when evaluating a set of rules/invariants, we should focus on their internal consistency and known use cases rather than on the details of the current (possibly broken) implementation.

*If* the proposed "no StoreEntry without a Transients entry" invariant is the best one we can come up with, but purgeMem() violates it, then we should fix purgeMem() or, at least, mark any bugs we cannot fix with XXXs and place them outside this PR scope. AFAICT, the only purgeMem()-related implementation problem is that MemStore::unlink() calls destroyMemObject(). However, that unconditional call could be a bug -- it is already marked with an XXX -- because the misnamed MemObject class stores more than the memory cache data (`data_hdr`, `memCache`, etc.): It also stores the list of the current Store clients (`clients` and `nclients`), part of the swapout state (`swapout`), a Transients lock (`xitTable`), `smpCollapsed` marker, `deferredReads`, etc.

Reply:
>  the only purgeMem()-related implementation problem is that MemStore::unlink() calls destroyMemObject().

What about other purgeMem() contexts, e.g., Controller::handleIdleEntry(), where it is called when  keepInLocalMemory is 'false'? Looking at cumbersome conditions within  Controller::dereferenceIdle(), I can't claim for sure, that the StoreEntry is always destroyed there, whereas it is always disconnected from Transients. If we follow your rule "no StoreEntry without a Transients entry", we would probably need to remove this complex handleIdleEntry() logic, including Controller::dereferenceIdle(), and simply destroy StoreEntry, if keepInLocalMemory is false.

Reply:
> I think we must guarantee that Transients::unlinkByKeyIfFound() is not called when there is a local StoreEntry object.

Or, equivalently, if Transients table exists, for each local StoreEntry hasTransients() must be true. destroyMemObject() disconnects Transients from the StoreEntry. I see two scenarios to consider:

1. Transients was disconnected while deleting StoreEntry, i.e., destroyStoreEntry() is called. In this case we firstly disconnect it from Transients and secondly remove it from store_table. Is it possible that anybody attaches to the already disconnected, but not yet deleted entry? Looks like impossible that this code became interrupted.

2. Transients was disconnected while purging memory (handleIdleEntry(), discussed above, or storeGetMemSpace() methods). As I noted above, there could be cases when such memobj-less StoreEntries keep existing in the local store_table. What is the purpose of leaving such half-baked  local entries? Optimization or anything else? If it is a bug, it should be fixed and the invariant becomes correct. If not, the invariant does not work.
Also, one thing is known that such transient-less and  memobj-less entries should be also unlocked, meaning that there are no clients attached to it. If this consideration is correct, the invariant should be valid without any additional fixes: we should not worry about such 'stand-alone' entries, because they stay apart from collapsing, notifying, etc.

Reply:
FYI: since we are dealing with a multi-process/thread environment any process may cause a kernel interrupt and do something with the shared parts unless the Transients disconnect part of the sequence is wrapped in proper atomic locks and the reader/writer counts checked inside the atomic lock (our StoreEntry locking does not count as thread-safety).

[ I have not checked the code to see if that is present, but you dont mention it as the reason for being "impossible", so though I should mention it. ]

Reply:
@yadij, AFAICT, when talking about "impossible" and "code interrupts" in item 1, Eduard is actually talking about things happening inside one Squid process. Here, "code interrupts" means AsyncCalls and such rather than process interrupts . This particular concern is unrelated to SMP, and Eduard is right that it is impossible for a new Store client to attach to a being destroyed StoreEntry object (i.e., attach during destroyStoreEntry()).

Reply:
>  Eduard is actually talking about things happening inside one Squid process.

That is correct. My supposition was that if Squid worker is suspended by OS at the beginning of StoreEntry destruction, it will resume executing from that very point of code, finishing the destruction process. No other clients can attach during this [interrupted] destruction process.

Reply:
> Alex: it is impossible for a new Store client to attach to a being destroyed StoreEntry object (i.e., attach during destroyStoreEntry()).

> Eduard: No other clients can attach during destruction process

For the record, both statements above talk about necessarily local (i.e., same-process) clients attaching to the StoreEntry object (which is always local!). Store clients in other workers (i.e., "remote" clients) may attach to the (shared) Transients entry and/or other StoreEntry objects in their processes, but that is a different issue.